### PR TITLE
feat: Provide averages binning to improve RF accuracy for some datasets

### PR DESCRIPTION
**Pull request by ahuber21**
* Provide `binningStrategy` and `maxBins=0` options
* Updated test cases for RF with tigther log loss ratio constraints (1.55 -> 1.15)
* Refactored RF classes: putting class methods back to class body (remove them from global module scope)
* Added `pyproject.toml` with instructions for `black` - formatted the files I touched

Because of the formatting changes, (`black` and `CLRF` to `LF`), the diff is very large. The individual contributions are provided in dedicated commits which might be easier to review.

### CODE REVIEW

File: daal4py/sklearn/ensemble/_forest.py [hunk 1]
```diff
@@ -1,1138 +1,1246 @@
...
+                raise ValueError(
+                    "min_samples_leaf must be at least 1 "
+                    f"or in (0, 0.5], got {self.min_samples_leaf}"
+                )
+        else:  # float
+            if not 0.0 < self.min_samples_leaf <= 0.5:
+                raise ValueError(
+                    "min_samples_leaf must be at least 1 "
+                    f"or in (0, 0.5], got {self.min_samples_leaf}"
+                )
```
Comment (KulikovNikita) [inline daal4py/sklearn/ensemble/_forest.py:171]: Formatting can be reused

File: daal4py/sklearn/ensemble/_forest.py [hunk 2]
```diff
@@ -1,1138 +1,1246 @@
...
+            [
+                (hasattr(self, "daal_model_"), "oneDAL model was not trained."),
+                (not sp.issparse(X), "X is sparse. Sparse input is not supported."),
+            ]
+        )
+        if hasattr(self, "n_outputs_"):
+            _dal_ready = _patching_status.and_conditions(
+                [
+                    (
+                        self.n_outputs_ == 1,
```
Comment (KulikovNikita) [inline daal4py/sklearn/ensemble/_forest.py:484]: Should you compare with int or float?

File: daal4py/sklearn/ensemble/_forest.py [hunk 3]
```diff
@@ -1,1138 +1,1246 @@
...
+                return (
+                    max(1, int(np.sqrt(n_features))) if is_classification else n_features
+                )
+        if max_features == "sqrt":
+            return max(1, int(np.sqrt(n_features)))
+        if max_features == "log2":
+            return max(1, int(np.log2(n_features)))
+        allowed_string_values = (
+            '"sqrt" or "log2"'
+            if sklearn_check_version("1.3")
```
[the remark this answers is not in the document]
Comment (ahuber21) [reply daal4py/sklearn/ensemble/_forest.py:72]: It was added here and I think it's consistent
https://github.com/intel/scikit-learn-intelex/commit/a09ceef7175707099e4417ac674e786845a72262
@Alexsandruss maybe you could confirm?

File: daal4py/sklearn/ensemble/_forest.py (hunk 2 above)
Comment (ahuber21) [reply daal4py/sklearn/ensemble/_forest.py:484]: `n_outputs_` is an integer

File: daal4py/sklearn/ensemble/_forest.py (hunk 1 above)
Comment (ahuber21) [reply daal4py/sklearn/ensemble/_forest.py:171]: How do you mean that?

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+[tool.black]
+line-length = 90
+target-version = ['py37', 'py38', 'py39', 'py310', 'py311']
```
Comment (KulikovNikita) [inline pyproject.toml:3]:
```suggestion
target-version = ['py38', 'py39', 'py310', 'py311']
```

Comment (napetrov) [reply pyproject.toml:3]: we still does support python 3.7 at least now. probably would be dropping it later this year

**Comment by ahuber21** [comment]
> One PreCommit step is failed already. This particular step failure should be investigated.

It's because we were running with the wrong oneDAL version. We need to check
intel-ci.intel.com/edfaeffa-f08d-f1ec-8eba-a4bf010d0e2e

**Comment by Alexsandruss** [comment]
@Mergifyio rebase

**Comment by Alexsandruss** [comment]
http://intel-ci.intel.com/edfc081a-3397-f16e-aeed-a4bf010d0e2e

**Comment by Alexsandruss** [comment]
http://intel-ci.intel.com/edfc9356-5ed2-f116-8266-a4bf010d0e2e

**Comment by Alexsandruss** [comment]
I suggest to wait for https://github.com/intel/scikit-learn-intelex/pull/1313, so this PR will be tested with correct sklearn versions

**Review by Alexsandruss — APPROVED** [review]
